sklearn/utils/validation.py

```diff
@@ -1980,9 +1989,10 @@ def _check_sample_weight(
     dtype : dtype, default=None
         dtype of the validated `sample_weight`.
         If None, and the input `sample_weight` is an array, the dtype of the
-        input is preserved; otherwise an array with the default numpy dtype
-        is be allocated.  If `dtype` is not one of `float32`, `float64`,
-        `None`, the output will be of dtype `float64`.
+        input is preserved; otherwise an array with the default dtype
+        is allocated.  If `dtype` is not `None` or of the "real floating" kind,
+        the output will be the maximum precision floating-point dtype for the
+        array namespace.
```

Comment:
In would rather not do that ahead of time in tools for `sklearn.utils.validation` but instead delay such an upcast to the estimators code and only done if really needed.

Reply:
Instead we should use the default float dtype of the xp namespace. For PyTorch, this is float32.

But most of the time, scikit-learn estimators should call `_check_sample_weight` with an explicit `dtype=` value, typically `dtype=X.dtype`.

Reply:
The implemented language matches the existing non-array-api behavior, which was to coerce the data type to `float64` when an explicit, non-floating-point dtype is provided. I'm not opposed to changing the behavior to only coerce to float, however. :+1:

Reply:
`float` does not exist in Array API parlance, and is a deprecated dtype for NumPy as well.

But there is indeed a default floating point dtype that is specific to each each Array API namespace:

https://data-apis.org/array-api/latest/API_specification/data_types.html#default-data-types.

As far as I know, there is no declarative way specified in the standard to discover the default floating dtype of a given namespace. But it's possible to inspect it using the following rather cheap construct: `xp.asarray(0.0).dtype`.